src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -2858,95 +2858,437 @@ private static CommandBaseAst GetPreviousPipelineCommand(CommandAst commandAst)
+            /// </summary>
+            internal ITypeName LastConstraint;
+            
+            /// <summary>
+            /// The last ast that assigned a value to the variable. This determines the value of the variable unless a type constraint has been applied.
+            /// </summary>
+            internal Ast LastAssignment;
+            
+            /// <summary>
+            /// The inferred type from the most recent assignment. This is only used for stream redirections to variables, or the special OutVariable common parameters.
+            /// </summary>
+            internal PSTypeName LastAssignmentType;
+            
+            /// <summary>
+            /// Whether or not the types from the last assignment should be enumerated.
+            /// For assignments made by the PipelineVariable parameter or the foreach statement.
+            /// </summary>
+            internal bool EnumerateAssignment;
+            
+            /// <summary>
+            /// Whether or not the last assignment was via command redirection.
+            /// </summary>
+            internal bool RedirectionAssignment;
+            internal int StopSearchOffset;
+            private int LastAssignmentOffset = -1;
+
+            private void SetLastAssignment(Ast ast, bool enumerate = false, bool redirectionAssignment = false)
+            {
+                if (LastAssignmentOffset < ast.Extent.StartOffset)
+                {
+                    LastAssignment = ast;
+                    EnumerateAssignment = enumerate;
+                    RedirectionAssignment = redirectionAssignment;
+                    LastAssignmentOffset = ast.Extent.StartOffset;
+                }
+            }
+
+            private void SetLastAssignmentType(PSTypeName typeName, int assignmentOffset)
+            {
+                if (LastAssignmentOffset < assignmentOffset)
+                {
+                    LastAssignment = null;
+                    LastAssignmentType = typeName;
+                    LastAssignmentOffset = assignmentOffset;
+                }
+            }
+
+            private bool AssignsToTargetVar(VariableExpressionAst foundVar)
+            {
+                if (!foundVar.VariablePath.UnqualifiedPath.EqualsOrdinalIgnoreCase(VariableTarget.VariablePath.UnqualifiedPath))
+                {
+                    return false;
+                }
 
-        public static IEnumerable<MethodInfo> GetGetterProperty(this Type type, string propertyName)
-        {
-            var res = new List<MethodInfo>();
-            foreach (var m in type.GetMethods(BindingFlags.Public | BindingFlags.Instance))
+                int scopeIndex = foundVar.VariablePath.UserPath.IndexOf(':');
+                string scopeName = scopeIndex == -1 ? string.Empty : foundVar.VariablePath.UserPath.Remove(scopeIndex);
+                return AssignsToTargetScope(scopeName);
+            }
+
+            private bool AssignsToTargetVar(string userPath)
```

Comment:
Can this method be reused in previous method?

Reply:
In theory it could, but it would be doing some additional pointless work when we have a VariableExpressionAst, which would be the case most of the time.